spring-app/src/main/java/pl/edu/agh/cqm/service/impl/ThroughputServiceImpl.java

```diff
@@ -98,33 +134,74 @@ private boolean parseDNS(DnsPacket dnsPacket) {
-                    .getAnswers()
-                    .stream()
+                    .getAnswers());
+            List<DnsRDataA> dnsARecords = dnsRecords.stream()
                     .filter(r -> r.getDataType().valueAsString().equals("1"))
-                    .collect(Collectors.toList());
-            if (dnsRecords.size() == 0) {
+                    .map(r -> ((DnsRDataA) r.getRData())).collect(Collectors.toList());
+            if(dnsARecords.size() == 0){
+                return false;
+            }
+            String ip = dnsARecords.get(0).getAddress().getHostAddress();
+            if(monitoredIPs.contains(ip)){
+                //logger.warn("this shouldn't happen; IP already in hashset. {}:{}:{}", url.cdn, url.name, ip);
                 return false;
             }
-            dnsRecords.stream()
-                    .map(r -> ((DnsRDataA) r.getRData()).getAddress().getHostAddress())
-                    .forEach(urlData::tryAdd);
-            logger.debug("found dns response for url {} : {} : {}", urlData.url.getAddress(), dnsPacket.getHeader().getQuestions().get(0).getQName().toString(), urlData.ips.get(0));
+            monitoredIPs.add(url.ip);
+            url.ip = dnsARecords.get(0).getAddress().getHostAddress();
+
+            logger.debug("found URL dns response cdn:{};  query:{};  ip:{}; url:{};", url.cdn, dnsPacket.getHeader().getQuestions().get(0).getQName().toString(), url.ip, url.name);
+
             return true;
+        }
+        for (String cdn : cdns) {
 
+                if (!dnsPacket.getHeader().getQuestions().get(0).getQName().toString().contains(cdn)) {
+                    continue;
+                }
+                List<DnsResourceRecord> dnsRecords = new ArrayList<>(dnsPacket
+                        .getHeader()
+                        .getAnswers());
+                if (dnsRecords.size() == 0) {
+                    return false;
+                }
+                List<DnsRDataA> dnsARecords = dnsRecords.stream()
+                        .filter(r -> r.getDataType().valueAsString().equals("1"))
+                        .map(r -> ((DnsRDataA) r.getRData())).collect(Collectors.toList());
+
+                if(dnsARecords.size() == 0){
+                    logger.debug("DNS A RECORDS 0");
+                    return false;
+                }
+                for (DnsRDataA record : dnsARecords) {
+                    if(monitoredIPs.contains(record.getAddress().getHostAddress())) return false;
+
+                    URLData urlData = new URLData();
+                    urlData.ip = record.getAddress().getHostAddress();
+                    urlData.cdn = cdn;
+
+                    urlsData.add(urlData);
+                    monitoredIPs.add(urlData.ip);
+                    logger.debug("found CDN dns response cdn:{};  query:{};  ip:{}; url:{};", cdn, dnsPacket.getHeader().getQuestions().get(0).getQName().toString(), urlData.ip, urlData.name);
+                    break;
+                }
```

Comment:
Is there a reason for a for loop here? There always is `return false` or `break` in the first iteration.